datafusion/src/physical_plan/functions.rs

```diff
@@ -468,7 +468,15 @@ pub fn return_type(
         | BuiltinScalarFunction::Sin
         | BuiltinScalarFunction::Sqrt
         | BuiltinScalarFunction::Tan
-        | BuiltinScalarFunction::Trunc => Ok(DataType::Float64),
+        | BuiltinScalarFunction::Trunc => {
+            if arg_types.len() <= 0 {
+                return Err(DataFusionError::Internal(String::from(
+                    "Builtin Scalar Function does not support empty arguments",
+                )));
+            }
+            // return the same type as the input argument
+            return Ok(arg_types.get(0).unwrap().clone());
```

Comment:
Yes, the code cannot work because of type coercion rules. I checked the rules, it really supports int as input args type. Thus here we cannot return an error when the input type is not Float32 or Float64.

So I kept the default return type, but add a float32 match arm as a special case.

And I also changed the behavior of ColumnarValue::Scalar  to be the same as ColumnarValue::Array(array). The test case passed this time.  >_<